### ISSUE TITLE
DeepCopier does shallow copy

### ISSUE THREAD

**Issue by sebastianburckhardt** [dotnet/orleans#1278]
Perhaps I am misunderstanding how this is intended to work... but I am using a simple grain state definition

```csharp
[Serializable]
    public class SampleGrainState
    {
        public List<string> Messages { get; set; }

        public SampleGrainState()
        {
            Messages = new List<string>();
        }
    }
```

but in the generated code, the deep copy does only a shallow copy

```csharp
public static global::System.Object DeepCopier(global::System.Object original)
        {
            global::QueuedGrainSample.Grains.SampleGrainState input = ((global::QueuedGrainSample.Grains.SampleGrainState)original);
            global::QueuedGrainSample.Grains.SampleGrainState result = new global::QueuedGrainSample.Grains.SampleGrainState();
            result.@Messages = input.@Messages;
            global::Orleans.@Serialization.@SerializationContext.@Current.@RecordObject(original, result);
            return result;
        }
```

Is this is intentionally designed to be this way? That seems confusing -- I would have expected the deep copy to be semantically equivalent to serialization followed by deserialization. What is the right way to make a deep copy efficiently?

**Comment by gabikliot** [comment]
Sebastian, of course it must truly deep copy. Are you sure that's the generated code? If yes, looks like a bug to me.

**Comment by gabikliot** [comment]
@sebastianburckhardt , I was able to validate that the serialization code is indeed like you wrote.
It is definitely a bug.
I think the bug is on this line: https://github.com/dotnet/orleans/blob/master/src/OrleansCodeGenerator/SerializerGenerator.cs#L728
We should be deep copying the field, unless it is shallow copiable, which `List` is not.

I am very confident it used to work correctly in the past, so the bug is pretty new.

**Comment by sebastianburckhardt** [comment]
I will try to fix this today.

**Comment by sebastianburckhardt** [comment]
What is the process for debugging code generation? Since it runs during compilation, I can't really set a breakpoint.

**Comment by sergeybykov** [comment]
I usually configure Debug section of the ClientGenerator project in VS to run with a command line that I take from build output, e.g. "@obj\Debug\TestGrainInterfaces.codegen.args.txt" for the TestGrainInterfaces project, and set Working directory to that of the target project, i.e. src\TestGrainInterfaces\ in this example.

**Comment by gabikliot** [comment]
Fixed via #1280.
Thank you @sebastianburckhardt !
